File: InstrumentUtilityDotNet/SpectrumAnalyzerManager/Agilent_856x.cs
using System;

namespace InstrumentUtilityDotNet.SpectrumAnalyzerManager
{
    public class Agilent_856x :  ISpectrumAnalyzer
    {
        /// <summary>
        /// 获取设备ID号 
        /// </summary>
        public override string GetID()
        {
            string sendMsg = "ID?;";
            try
            {
                return base.WriteAndReadString(sendMsg);
            }
            catch (Exception ex)
            {
                throw (ex);
            }
        }
        /// <summary>
        /// 初始化仪表参数
        /// </summary>
        /// <returns></returns>
        public override bool Reset()
        {
            string sendMsg = "IP;";
            try
            {
                return base.WriteString(sendMsg);
            }
            catch (Exception ex)
            {
                throw (ex);
            }
        }



        /// <summary>
        /// 获取中心频率  
        /// </summary>
        /// <returns></returns>
        public override double GetCenterFreq()
        {
            string recvMsg = null;
            try
            {
                recvMsg = base.WriteAndReadString("CF?;");
                if (recvMsg != "")
                    return Convert.ToDouble(recvMsg);
                else
                    return 0;
            }
            catch (Exception ex)
            {
                throw (ex);
            }
        }


        /// <summary>
        /// 获取MKA  峰值电平
        /// </summary>
        /// <returns></returns>
        public override double GetMKA()
        {
            string recvMsg = null;
            try
            {
                recvMsg = base.WriteAndReadString("MKA?;");
                double x = -999;
                bool res = double.TryParse(recvMsg, out x);
                if (res)
                    return x;
                else
                    return -999;
            }
            catch (Exception ex)
            {
                throw (ex);
            }

        }

        /// <summary>
        /// 设置中心频率  
        /// </summary>
        /// <param name="value">值</param>
        /// <param name="unit">频率单位</param>
        /// <returns></returns>
        public override bool SetCenterFreq(double value, FrequencyUnit unit)
        {
            string sendMsg = "CF  " + value;
            switch (unit)
            {
                case FrequencyUnit.Hz:
                    sendMsg += "Hz";
                    break;
                case FrequencyUnit.KHz:
                    sendMsg += "KHz";
                    break;
                case FrequencyUnit.MHz:
                    sendMsg += "MHz";
                    break;
                case FrequencyUnit.GHz:
                    sendMsg += "GHz";
                    break;
            }

            try
            {
                return base.WriteString(sendMsg);
            }
            catch (Exception ex)
            {
                throw (ex);
            }
        }

        /// <summary>
        /// 设置RBW  
        /// </summary>
        /// <param name="isAuto">自动/手动RBW</param>
        /// <param name="value">分辨率带宽，单位KHz</param>
        /// <returns></returns>
        public override bool SetRBW(bool isAuto, double value = 0)
        {
            string sendMsg = null;
            try
            {
                if (isAuto)
                {
                    sendMsg = "BAND:AUTO ON";
                    return base.WriteString(sendMsg);
                }
                else
                {
                    sendMsg = "BAND:AUTO OFF;RB " + value + "kHz;";
                    return base.WriteString(sendMsg);
                }
            }
            catch (Exception ex)
            {
                throw (ex);
            }
        }

      

        /// <summary>
        /// 设置参考电平 
        /// </summary>
        /// <param name="value">参考电平,单位dBm</param>
        /// <returns></returns>
        public override bool SetRefLevel(double value)
        {
            string sendMsg = "RL " + value + "dBm;";
            try
            {
                return base.WriteString(sendMsg);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                return false;
            }
        }

        /// <summary>
        /// 激活标记并搜索峰值  
        /// </summary>
        /// <returns></returns>
        public override bool MarkPeak()
        {
            string sendMsg = "TS;MKPK HI;";
            try
            {
                return base.WriteString(sendMsg);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                return false;
            }
        }

        /// <summary>
        /// 设置衰减
        /// </summary>
        /// <param name="isAuto">自动/手动衰减</param>
        /// <param name="value">衰减值 单位DB</param>
        /// <returns></returns>
        public override bool SetAttenuation(bool isAuto,double value)
        {
            string sendMsg = null;
            if (isAuto)
            {
                sendMsg = "AT AUTO";
            }
            else
            {
                sendMsg = "AT MANUAL;AT  " + value + "DB";
            }
            try
            {
                return base.WriteString(sendMsg);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                return false;
            }

        }

    

        /// <summary>
        /// 设置SPAN
        /// </summary>
        /// <param name="value">带宽</param>
        /// <param name="unit">单位</param>
        /// <returns></returns>
        public override bool SetSpan(double value,FrequencyUnit unit)
        {
            string sendMsg = "SP " + value + unit.ToString();
            try
            {
                return base.WriteString(sendMsg);
 
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                return false;
            }
        }

        /// <summary>
        /// 线损设置 REF LVL OFFSET
        /// </summary>
        /// <param name="value"></param>
        /// <returns></returns>
        public override bool SetRefOffset(double value)
        {
            string sendMsg = "ROFFSET " + value +"dB";
            try
            {
                return base.WriteString(sendMsg);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                return false;
            }
        }

    }
}

File: InstrumentUtility/InstrumentUtility/Agilent_856x.cpp

#include "stdafx.h"
#include "Agilent_856x.h"
#include "stdlib.h"
#include <string>

using namespace std;

Agilent_856x::Agilent_856x()
{

}

Agilent_856x::~Agilent_856x()
{

}

char* Agilent_856x::GetID()
{
	char* sendMsg = "ID?;";
	char* recvMsg="";
	try
	{
		bool res = WriteAndReadString(sendMsg, recvMsg);
 
		return recvMsg;
	}
	catch (char *str)
	{
		throw(str);
	}
	return nullptr;
}

bool Agilent_856x::Reset()
{
	char* sendMsg = "IP;";
	try
	{
		WriteString(sendMsg);
		return true;
	}
	catch (char *str)
	{
		throw(str);
	}
	return false;
}

double Agilent_856x::GetCenterFreq()
{
	char* sendMsg = "CF?;";
	char* recvMsg="";
	try
	{
		bool res = WriteAndReadString(sendMsg, recvMsg);
		return atof(recvMsg);
	}
	catch (char *str)
	{
		throw(str);
	}
	return -1;
}

double Agilent_856x::GetMKA()
{
	char* sendMsg = "MKA?;";
	char* recvMsg = "";
	try
	{
		bool res = WriteAndReadString(sendMsg, recvMsg);
		return atof(recvMsg);
	}
	catch (char *str)
	{
		throw(str);
	}
	return -999;
}

bool Agilent_856x::SetCenterFreq(FrequencyUnit unit, double value)
{
	string sendMsg = "CF  " ;
	sendMsg += value;
	switch (unit)
	{
	case FrequencyUnit::Hz:
		sendMsg += "Hz";
		break;
	case FrequencyUnit::KHz:
		sendMsg += "KHz";
		break;
	case FrequencyUnit::MHz:
		sendMsg += "MHz";
		break;
	case FrequencyUnit::GHz:
		sendMsg += "GHz";
		break;
	}
	try
	{
		WriteString(const_cast<char*>(sendMsg.data()));
		return true;
	}
	catch (char *str)
	{
		throw(str);
	}
	return false;
}

bool Agilent_856x::SetRBW(bool isAuto, double value)
{
	string sendMsg = "BAND:AUTO ";
	bool res = false;
	try
	{
		if (isAuto)
		{
			sendMsg += "ON";
		    res = WriteString(const_cast<char*>(sendMsg.data()));
			return res;
		}
		else
		{
			sendMsg += "OFF";
		    res = WriteString(const_cast<char*>(sendMsg.data()));
 
			char* buf = new char[sizeof(value) + 8];   //设置RBW值
			printf_s(buf, "RB %.2f kHz;", value);
			res = WriteString(buf);
			return res;
		}
	}
	catch (char* str)
	{
		throw(str);
	}
	return false;
}

bool Agilent_856x::SetRefLevel(double value)
{
	char* sendMsg = new char[sizeof(value) + 8];   //设置RBW值
	bool res = false;
	try
	{
		printf_s(sendMsg, "RL %f dBm;", value);
		res = WriteString(sendMsg);
		return res;
	}
	catch (char* str)
	{
		throw(str);
	}
	return false;
}


bool Agilent_856x::MarkPeak()
{
	char* sendMsg = "TS;MKPK HI;";   
	bool res = false;
	try
	{
		res = WriteString(sendMsg);
		return res;
	}
	catch (char* str)
	{
		throw(str);
	}
	return false;
}

bool Agilent_856x::SetAttenuation(bool isAuto, double value)
{
	char* sendMsg = nullptr ;
	bool res = false;
	try
	{
		if (isAuto)
		{
			sendMsg = "AT AUTO";
			res = WriteString(sendMsg);
			return res;
		}
		else
		{
			sendMsg = new char[sizeof(value) + 8];   
			printf_s(sendMsg, "AT MANUAL;AT f% DB", value);
			res = WriteString(sendMsg);
			return res;
		}
	}
	catch (char* str)
	{
		throw(str);
	}
	return false;
}

bool Agilent_856x::SetSpan(double value, FrequencyUnit unit)
{
	string sendMsg = "SP ";
	sendMsg += value;
	switch (unit)
	{
	case FrequencyUnit::Hz:
		sendMsg += "Hz";
		break;
	case FrequencyUnit::KHz:
		sendMsg += "KHz";
		break;
	case FrequencyUnit::MHz:
		sendMsg += "MHz";
		break;
	case FrequencyUnit::GHz:
		sendMsg += "GHz";
		break;
	}
	try
	{
		WriteString(const_cast<char*>(sendMsg.data()));
		return true;
	}
	catch (char *str)
	{
		throw(str);
	}
	return false;
}

 
	
